src/telemetry/downlink_queue/downlink_queue.c

```diff
@@ -1,14 +1,172 @@
+ * of a highest priority.
+ *
+ * @TODO Question - In header file, do we define every prototype that we need
+ *       to use, or just the ones that are used outside, and the rest in the
+ *       file locally?
+ * @TODO Question - Do we define MAX_LENGTH for the queue? If so, is it max
+ *       size of the data in it, or max number of nodes?
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <pthread.h>
+
 #include "global_utils.h"
 
-int init_downlink_queue( void ){
+#include "downlink_queue.h"
+
+pthread_mutex_t downlink_mutex = PTHREAD_MUTEX_INITIALIZER;
+pthread_cond_t queue_non_empty_cond = PTHREAD_COND_INITIALIZER;
+
+int init_downlink_queue(void) {
+    return SUCCESS;
+}
+
+/**
+ * Node structure declaration.
+ */
+typedef struct node {
+    int data;           // Data to be send as a telemetry.
+    int priority;       // Lower values indicate higher priority
+    struct node *next;  // Pointer to the node next on the list.
+
+} downlink_node;
+
+/**
+ * Function to create a new node
+ *
+ * @param d     Data to be sent.
+ * @param p     Priority of the data.
+ */
+downlink_node *new_node(int d, int p) {
+    downlink_node *temp = (downlink_node *) malloc(sizeof(downlink_node));
+    temp->data = d;
+    temp->priority = p;
+    temp->next = NULL;
+
+    return temp;
+}
+
+/**
+ * Function to check if the list is empty
+ *
+ * @param head  Pointer to the first node of the linked list.
+ * @return      1 if the list is empty, 0 if not.
+ */
+int is_empty(downlink_node **head) {
+    return (*head) == NULL;
+}
+
+int peek(downlink_node **head) {
+    if (!is_empty(head)) {
+        return (*head)->data;
+    } else {
+        return FAILURE;
+    }
+}
+
+/**
+ * Removes the element with the highest priority form the list and
+ * return it's data value.
+ *
+ * @param head  Pointer to the first node of the linked list.
+ * @return      Data from the popped node.
+ */
+int pop(downlink_node **head) {
+    pthread_mutex_lock(&downlink_mutex);
+
+    while (is_empty(head)) {
+        pthread_cond_wait(&queue_non_empty_cond, &downlink_mutex);
+    }
+
+    downlink_node *temp = *head;
+    int data = (*head)->data;
+    (*head) = (*head)->next;
+    free(temp);
+
+    pthread_mutex_unlock(&downlink_mutex);
+    return data;
+}
+
+/**
+ * Function to push node to the list according to its priority.
+ *
+ * @param head  Pointer to the first node of the linked list.
+ * @param d     Data to be sent.
+ * @param p     Priority of the data.
+ */
+void push(downlink_node **head, int d, int p) {
+    pthread_mutex_lock(&downlink_mutex);
+    downlink_node *start = (*head);
+
+    if (!is_empty(head)) {
+        // Create new node
+        downlink_node *temp = new_node(d, p);
+        // Special Case: The head of list has lesser
+        // priority than new node. So insert new
+        // node before head node and change head node.
+        if ((*head)->priority > p) {
```

Comment:
Should it not be like this?
```suggestion
        if ((*head)->priority < p) {
```

Reply:
The lower the `p`, the higher the priority. So `0` is the highest one. `head` is the existing node, so if it has lower priority (so higher number)  than `p` of the data being added, the new node is added as the first. Hence `if head.priority > p`.